From 22ab8a9bb3ce5561815edf5c8d40952e77be50d7 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Fri, 19 Apr 2024 10:41:53 +0200
Subject: [PATCH 0892/1695] product: add tfm-stm32mp2 product.

Since only stm32mp2 platform is targeted for scp support
in tfm , all the modules are product specific. The product definition is
based on optee stm32mp2 product.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8e40413dd155127474f600d265a52f03a3d3f065 product: add tfm-stm32mp2 product.]
Change-Id: I9db238d5b685d646e1e91527ec01b75fccecfbf6
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/SCP-firmware/+/383539
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/385195
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 .../product/tfm-stm32mp2/fw/CMakeLists.txt    |  29 +
 .../product/tfm-stm32mp2/fw/Firmware.cmake    |  54 ++
 .../tfm-stm32mp2/fw/Toolchain-GNU.cmake       |  23 +
 .../product/tfm-stm32mp2/fw/config_all.c      | 583 ++++++++++++++++++
 .../product/tfm-stm32mp2/include/fmw_io.h     |  17 +
 .../product/tfm-stm32mp2/include/fmw_log.h    |  16 +
 .../tfm-stm32mp2/include/gnu/stubs-soft.h     |   0
 .../tfm-stm32mp2/include/scmi_agents.h        |  13 +
 .../module/tfm_clock/CMakeLists.txt           |  25 +
 .../module/tfm_clock/Module.cmake             |   9 +
 .../module/tfm_clock/include/mod_tfm_clock.h  |  32 +
 .../module/tfm_clock/src/mod_tfm_clock.c      | 369 +++++++++++
 .../module/tfm_console/CMakeLists.txt         |  19 +
 .../module/tfm_console/Module.cmake           |   9 +
 .../module/tfm_console/src/mod_tfm_console.c  |  93 +++
 .../module/tfm_mbx/CMakeLists.txt             |  27 +
 .../tfm-stm32mp2/module/tfm_mbx/Module.cmake  |  10 +
 .../module/tfm_mbx/include/mod_tfm_mbx.h      |  67 ++
 .../module/tfm_mbx/src/mod_tfm_mbx.c          | 258 ++++++++
 .../module/tfm_regu_consumer/CMakeLists.txt   |  25 +
 .../module/tfm_regu_consumer/Module.cmake     |  10 +
 .../include/mod_tfm_regu_consumer.h           |  27 +
 .../src/mod_tfm_regu_consumer.c               | 294 +++++++++
 .../module/tfm_reset/CMakeLists.txt           |  24 +
 .../module/tfm_reset/Module.cmake             |  10 +
 .../module/tfm_reset/include/mod_tfm_reset.h  |  23 +
 .../module/tfm_reset/src/mod_tfm_reset.c      | 172 ++++++
 27 files changed, 2238 insertions(+)
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/CMakeLists.txt
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Firmware.cmake
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Toolchain-GNU.cmake
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/config_all.c
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_io.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_log.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/gnu/stubs-soft.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/scmi_agents.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/CMakeLists.txt
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/Module.cmake
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/include/mod_tfm_clock.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/src/mod_tfm_clock.c
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/CMakeLists.txt
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/Module.cmake
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/src/mod_tfm_console.c
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/CMakeLists.txt
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/Module.cmake
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/include/mod_tfm_mbx.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/CMakeLists.txt
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/Module.cmake
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/include/mod_tfm_regu_consumer.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/src/mod_tfm_regu_consumer.c
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/CMakeLists.txt
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/Module.cmake
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/include/mod_tfm_reset.h
 create mode 100644 core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/src/mod_tfm_reset.c

diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/CMakeLists.txt b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/CMakeLists.txt
new file mode 100644
index 000000000..040e655b8
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/CMakeLists.txt
@@ -0,0 +1,29 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+#
+# Create the firmware target.
+#
+
+add_library(${SCP_FIRMWARE_TARGET})
+
+# cmake-lint: disable=E1122
+
+target_include_directories(
+    ${SCP_FIRMWARE_TARGET}
+    PUBLIC  ${CMAKE_CURRENT_SOURCE_DIR}/../include
+            ${SCP_TFM_PLAT_DIR}/native_driver/include
+            ${SCP_TFM_PLAT_DIR}/../devicetree/include
+            ${SCP_TFM_PLAT_DIR}/native_driver
+            ${CMAKE_BINARY_DIR}/generated/devicetree/platform_s
+            ${STM_SOC_DIR}/secure_fw/partitions/scp
+)
+
+target_sources(
+    ${SCP_FIRMWARE_TARGET}
+    PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/config_all.c"
+)
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Firmware.cmake b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Firmware.cmake
new file mode 100644
index 000000000..4dade6080
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Firmware.cmake
@@ -0,0 +1,54 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# Configure the build system.
+#
+
+set(SCP_FIRMWARE "scmi-fw")
+
+set(SCP_FIRMWARE_TARGET "scmi-fw")
+
+set(SCP_TOOLCHAIN_INIT "GNU")
+
+set(SCP_ARCHITECTURE "tfm")
+
+set(CMAKE_BUILD_TYPE "Release")
+
+set(SCP_ENABLE_NOTIFICATIONS_INIT FALSE)
+
+set(SCP_ENABLE_SCMI_NOTIFICATIONS_INIT FALSE)
+
+set(SCP_ENABLE_SCMI_SENSOR_EVENTS_INIT FALSE)
+
+set(SCP_ENABLE_FAST_CHANNELS_INIT FALSE)
+
+set(SCP_ENABLE_SCMI_RESET_INIT TRUE)
+
+set(SCP_ENABLE_IPO_INIT FALSE)
+
+list(PREPEND SCP_MODULE_PATHS "${CMAKE_CURRENT_LIST_DIR}/../module/tfm_regu_consumer")
+list(PREPEND SCP_MODULE_PATHS "${CMAKE_CURRENT_LIST_DIR}/../module/tfm_clock")
+list(PREPEND SCP_MODULE_PATHS "${CMAKE_CURRENT_LIST_DIR}/../module/tfm_mbx")
+list(PREPEND SCP_MODULE_PATHS "${CMAKE_CURRENT_LIST_DIR}/../module/tfm_reset")
+list(PREPEND SCP_MODULE_PATHS "${CMAKE_CURRENT_LIST_DIR}/../module/tfm_console")
+
+# The order of the modules in the following list is the order in which the
+# modules are initialized, bound, started during the pre-runtime phase.
+# any change in the order will cause firmware initialization errors.
+
+list(APPEND SCP_MODULES "tfm-mbx")
+list(APPEND SCP_MODULES "msg-smt")
+list(APPEND SCP_MODULES "scmi")
+list(APPEND SCP_MODULES "tfm-clock")
+list(APPEND SCP_MODULES "clock")
+list(APPEND SCP_MODULES "scmi-clock")
+list(APPEND SCP_MODULES "tfm-regu-consumer")
+list(APPEND SCP_MODULES "voltage-domain")
+list(APPEND SCP_MODULES "scmi-voltage-domain")
+list(APPEND SCP_MODULES "tfm-reset")
+list(APPEND SCP_MODULES "reset-domain")
+list(APPEND SCP_MODULES "scmi-reset-domain")
+list(APPEND SCP_MODULES "tfm-console")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Toolchain-GNU.cmake b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Toolchain-GNU.cmake
new file mode 100644
index 000000000..cfa6d8cd7
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/Toolchain-GNU.cmake
@@ -0,0 +1,23 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+include_guard()
+
+set(CMAKE_SYSTEM_PROCESSOR "tfm")
+set(CMAKE_TOOLCHAIN_PREFIX ${CFG_CROSS_COMPILE})
+
+set(CMAKE_ASM_COMPILER_TARGET ${CFG_CROSS_COMPILE})
+set(CMAKE_C_COMPILER_TARGET ${CFG_CROSS_COMPILE})
+set(CMAKE_CXX_COMPILER_TARGET ${CFG_CROSS_COMPILE})
+
+include(
+    "${CMAKE_CURRENT_LIST_DIR}/../../../cmake/Toolchain/GNU-Baremetal.cmake")
+
+foreach(language IN ITEMS ASM C CXX)
+    string(APPEND CMAKE_${language}_FLAGS_INIT ${CFG_CFLAGS_TFM})
+endforeach()
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/config_all.c b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/config_all.c
new file mode 100644
index 000000000..3ff0ca2a0
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/fw/config_all.c
@@ -0,0 +1,583 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022-2023, Linaro Limited and Contributors. All rights reserved.
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <regulator.h>
+#include <fwk_assert.h>
+#include <fwk_element.h>
+#include <fwk_id.h>
+#include <fwk_log.h>
+#include <fwk_mm.h>
+#include <fwk_module.h>
+#include <fwk_module_idx.h>
+
+#include <mod_clock.h>
+#include <mod_msg_smt.h>
+#include <mod_tfm_clock.h>
+#include <mod_tfm_mbx.h>
+#include <mod_tfm_reset.h>
+#include <mod_reset_domain.h>
+#include <mod_scmi.h>
+#include <mod_scmi_clock.h>
+#include <mod_scmi_voltage_domain.h>
+#include <mod_tfm_regu_consumer.h>
+#include <mod_scmi_reset_domain.h>
+#include <scmi_agents.h>
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <arch_helpers.h>
+#include <scmi_agent_configuration.h>
+
+/* SCMI agent and services (channels) */
+static struct mod_scmi_agent *scmi_agent_table;
+static struct mod_scmi_config scmi_data;
+static struct fwk_element *scmi_service_elt;
+
+/* SCMI channel mailbox/shmem */
+static struct fwk_element *msg_smt_elt;
+static struct mod_msg_smt_channel_config *msg_smt_data;
+static struct fwk_element *tfm_mbx_elt;
+static struct mod_tfm_mbx_channel_config *tfm_mbx_data;
+
+#ifdef CFG_SCPFW_MOD_SCMI_CLOCK
+/* SCMI clock generic */
+static struct mod_scmi_clock_agent *scmi_clk_agent_tbl;
+#endif
+
+#ifdef CFG_SCPFW_MOD_CLOCK
+/* Clocks and tfm/clock, same number/indices. Elements and configuration data */
+static struct fwk_element *tfm_clock_elt;		/* Optee/clock elements */
+static struct mod_tfm_clock_config *tfm_clock_cfg;	/* Config data for tfm/clock elements */
+static struct fwk_element *clock_elt;			/* Clock elements */
+static struct mod_clock_dev_config *clock_data;		/* Config data for clock elements */
+#endif
+
+#ifdef CFG_SCPFW_MOD_RESET_DOMAIN
+/* SCMI reset domains and tfm reset controller */
+static struct mod_scmi_reset_domain_agent *scmi_reset_agent_tbl;
+static struct fwk_element *tfm_reset_elt;
+static struct mod_tfm_reset_dev_config *tfm_reset_data;
+static struct fwk_element *reset_elt;
+static struct mod_reset_domain_dev_config *reset_data;
+#endif
+
+#ifdef CFG_SCPFW_MOD_VOLTAGE_DOMAIN
+/* SCMI voltage domains and tfm regulators */
+static struct mod_scmi_voltd_agent *scmi_voltd_agent_tbl;
+static struct fwk_element *tfm_regu_elt;
+static struct mod_tfm_regu_consumer_dev_config *tfm_regu_data;
+static struct fwk_element *voltd_elt;
+static struct mod_voltd_dev_config *voltd_data;
+#endif
+
+/* Config data for scmi module */
+static const struct fwk_element *get_scmi_service_table(fwk_id_t module_id)
+{
+    return scmi_service_elt; /* scmi_service_elt filled during initialization */
+}
+
+struct fwk_module_config config_scmi = {
+    .data = (void *)&scmi_data, /* scmi_data filled during initialization */
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(get_scmi_service_table),
+};
+
+/* Config data for tfm_mbx module */
+static const struct fwk_element *tfm_mbx_get_element_table(fwk_id_t module_id)
+{
+    return (const struct fwk_element *)tfm_mbx_elt;
+}
+
+struct fwk_module_config config_tfm_mbx = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(tfm_mbx_get_element_table),
+};
+
+/* Config data for msg_smt module */
+static const struct fwk_element *msg_smt_get_element_table(fwk_id_t module_id)
+{
+    fwk_assert(fwk_id_get_module_idx(module_id) == FWK_MODULE_IDX_MSG_SMT);
+    return (const struct fwk_element *)msg_smt_elt;
+}
+
+struct fwk_module_config config_msg_smt = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(msg_smt_get_element_table),
+};
+
+/* Config data for scmi_clock, clock and tfm_clock modules */
+#ifdef CFG_SCPFW_MOD_SCMI_CLOCK
+struct fwk_module_config config_scmi_clock = {
+    .data = &((struct mod_scmi_clock_config){
+        .agent_table = NULL,			/* Allocated during initialization */
+        .agent_count = 0,			/* Set during initialization */
+    }),
+};
+#endif
+
+#ifdef CFG_SCPFW_MOD_CLOCK
+static const struct fwk_element *clock_get_element_table(fwk_id_t module_id)
+{
+    fwk_assert(fwk_id_get_module_idx(module_id) == FWK_MODULE_IDX_CLOCK);
+    return (const struct fwk_element *)clock_elt;
+}
+
+struct fwk_module_config config_clock = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(clock_get_element_table),
+};
+
+static const struct fwk_element *tfm_clock_get_element_table(fwk_id_t module_id)
+{
+    fwk_assert(fwk_id_get_module_idx(module_id) == FWK_MODULE_IDX_TFM_CLOCK);
+    return (const struct fwk_element *)tfm_clock_elt;
+}
+
+struct fwk_module_config config_tfm_clock = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(tfm_clock_get_element_table),
+};
+#endif
+
+/* Config data for scmi_reset_domain, reset_domain and tfm_reset modules */
+#ifdef CFG_SCPFW_MOD_RESET_DOMAIN
+struct fwk_module_config config_scmi_reset_domain = {
+    .data = &((struct mod_scmi_reset_domain_config){
+        .agent_table = NULL,			/* Allocated during initialization */
+        .agent_count = 0,			/* Set during initialization */
+    }),
+};
+
+static const struct fwk_element *reset_get_element_table(fwk_id_t module_id)
+{
+    fwk_assert(fwk_id_get_module_idx(module_id) == FWK_MODULE_IDX_RESET_DOMAIN);
+    return (const struct fwk_element *)reset_elt;
+}
+
+struct fwk_module_config config_reset_domain = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(reset_get_element_table),
+};
+
+static const struct fwk_element *tfm_reset_get_element_table(fwk_id_t module_id)
+{
+    fwk_assert(fwk_id_get_module_idx(module_id) == FWK_MODULE_IDX_TFM_RESET);
+    return (const struct fwk_element *)tfm_reset_elt;
+}
+
+struct fwk_module_config config_tfm_reset = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(tfm_reset_get_element_table),
+};
+#endif
+
+#ifdef CFG_SCPFW_MOD_VOLTAGE_DOMAIN
+/* Config data for scmi_voltage_domain, voltage_domain and tfm_regu modules */
+struct fwk_module_config config_scmi_voltage_domain = {
+    .data = &((struct mod_scmi_voltd_config){
+        .agent_table = NULL,			/* Allocated during initialization */
+        .agent_count = 0,			/* Set during initialization */
+    }),
+};
+
+static const struct fwk_element *voltd_get_element_table(fwk_id_t module_id)
+{
+    fwk_assert(fwk_id_get_module_idx(module_id) == FWK_MODULE_IDX_VOLTAGE_DOMAIN);
+    return (const struct fwk_element *)voltd_elt;
+}
+
+struct fwk_module_config config_voltage_domain = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(voltd_get_element_table),
+};
+
+static const struct fwk_element *tfm_regu_consumer_get_element_table(fwk_id_t module_id)
+{
+    fwk_assert(fwk_id_get_module_idx(module_id) == FWK_MODULE_IDX_TFM_REGU_CONSUMER);
+    return (const struct fwk_element *)tfm_regu_elt;
+}
+
+struct fwk_module_config config_tfm_regu_consumer = {
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(tfm_regu_consumer_get_element_table),
+};
+#endif
+
+/*
+ * Indices state when applying agents configuration
+ * @channel_count: Number of channels (mailbox/shmem links) used
+ * @clock_index: Current index for clock and tfm/clock (same indices)
+ * @clock_count: Number of clocks (also number of tfm/clocks)
+ * @reset_index: Current index for reset controller and tfm/reset
+ * @reset_count: Number of reset controller (tfm/reset) instances
+ * @regu_index: Current index for voltd and tfm/regulator
+ * @regu_count: Number of voltd (tfm/regulator) instances
+ */
+struct scpfw_resource_counter {
+    size_t channel_count;
+    size_t clock_index;
+    size_t clock_count;
+    size_t reset_index;
+    size_t reset_count;
+    size_t regu_index;
+    size_t regu_count;
+} scpfw_resource_counter;
+
+/*
+ * Count once for all the several instances and allocate global resources.
+ * Global resources are clock, tfm/clock, reset, tfm/reset, regu,
+ * tfm/regu, psu, tfm/psu, dvfs, perfd, ...;
+ */
+static void count_resources(struct scpfw_config *cfg)
+{
+    size_t i, j;
+
+    for (i = 0; i < cfg->agent_count; i++) {
+        struct scpfw_agent_config *agent_cfg = cfg->agent_config + i;
+
+        scpfw_resource_counter.channel_count += agent_cfg->channel_count;
+
+        for (j = 0; j < agent_cfg->channel_count; j++) {
+            struct scpfw_channel_config *channel_cfg = agent_cfg->channel_config + j;
+
+            /* Clocks for scmi_clock */
+            scpfw_resource_counter.clock_count += channel_cfg->clock_count;
+            /* Reset for smci_reset only */
+            scpfw_resource_counter.reset_count += channel_cfg->reset_count;
+            /* Regulators for smci_voltage_domain only */
+            scpfw_resource_counter.regu_count += channel_cfg->voltd_count;
+        }
+    }
+
+#ifndef CFG_SCPFW_MOD_CLOCK
+    fwk_assert(!scpfw_resource_counter.clock_count);
+#endif
+#ifndef CFG_SCPFW_MOD_RESET_DOMAIN
+    fwk_assert(!scpfw_resource_counter.reset_count);
+#endif
+#ifndef CFG_SCPFW_MOD_VOLTAGE_DOMAIN
+    fwk_assert(!scpfw_resource_counter.regu_count);
+#endif
+}
+
+/*
+ * Allocate all tables that may be needed. An optimized implementation would
+ * allocate a single piece of memory and set the pointers accordingly.
+ * */
+static void allocate_global_resources(struct scpfw_config *cfg)
+{
+    struct mod_scmi_reset_domain_config *scmi_reset_config __maybe_unused;
+    struct mod_scmi_voltd_config *scmi_voltd_config __maybe_unused;
+    struct mod_scmi_clock_config *scmi_clock_config __maybe_unused;
+    /* @cfg does not consider agent #0: ID reserved for the platform/server */
+    size_t __maybe_unused scmi_agent_count = cfg->agent_count + 1;
+
+#ifdef CFG_SCPFW_MOD_SCMI_CLOCK
+    /* SCMI clock domains resources */
+    scmi_clk_agent_tbl = fwk_mm_calloc(scmi_agent_count,
+                                       sizeof(*scmi_clk_agent_tbl));
+    scmi_clock_config = (void *)config_scmi_clock.data;
+    scmi_clock_config->agent_table = scmi_clk_agent_tbl;
+    scmi_clock_config->agent_count = scmi_agent_count;
+#endif
+
+#ifdef CFG_SCPFW_MOD_CLOCK
+    /* Clock domains resources */
+    tfm_clock_cfg = fwk_mm_calloc(scpfw_resource_counter.clock_count,
+                                    sizeof(*tfm_clock_cfg));
+    tfm_clock_elt = fwk_mm_calloc(scpfw_resource_counter.clock_count + 1,
+                                    sizeof(*tfm_clock_elt));
+
+    clock_data = fwk_mm_calloc(scpfw_resource_counter.clock_count,
+                               sizeof(*clock_data));
+    clock_elt = fwk_mm_calloc(scpfw_resource_counter.clock_count + 1,
+                              sizeof(*clock_elt));
+#endif
+
+#ifdef CFG_SCPFW_MOD_RESET_DOMAIN
+    /* SCMI reset domains resources */
+    scmi_reset_agent_tbl = fwk_mm_calloc(scmi_agent_count,
+                                         sizeof(*scmi_reset_agent_tbl));
+    scmi_reset_config = (void *)config_scmi_reset_domain.data;
+    scmi_reset_config->agent_table = scmi_reset_agent_tbl;
+    scmi_reset_config->agent_count = scmi_agent_count;
+
+    tfm_reset_data = fwk_mm_calloc(scpfw_resource_counter.reset_count,
+                                     sizeof(*tfm_reset_data));
+    tfm_reset_elt = fwk_mm_calloc(scpfw_resource_counter.reset_count + 1,
+                                    sizeof(*tfm_reset_elt));
+
+    reset_data = fwk_mm_calloc(scpfw_resource_counter.reset_count,
+                               sizeof(*reset_data));
+    reset_elt = fwk_mm_calloc(scpfw_resource_counter.reset_count + 1,
+                              sizeof(*reset_elt));
+#endif
+
+#ifdef CFG_SCPFW_MOD_VOLTAGE_DOMAIN
+    /* SCMI voltage domains resources */
+    scmi_voltd_agent_tbl = fwk_mm_calloc(scmi_agent_count,
+                                         sizeof(*scmi_voltd_agent_tbl));
+    scmi_voltd_config = (void *)config_scmi_voltage_domain.data;
+    scmi_voltd_config->agent_table = scmi_voltd_agent_tbl;
+    scmi_voltd_config->agent_count = scmi_agent_count;
+
+    tfm_regu_data = fwk_mm_calloc(scpfw_resource_counter.regu_count,
+                                    sizeof(*tfm_regu_data));
+    tfm_regu_elt = fwk_mm_calloc(scpfw_resource_counter.regu_count + 1,
+                                    sizeof(*tfm_regu_elt));
+
+    voltd_data = fwk_mm_calloc(scpfw_resource_counter.regu_count,
+                               sizeof(*voltd_data));
+    voltd_elt = fwk_mm_calloc(scpfw_resource_counter.regu_count + 1,
+                              sizeof(*voltd_elt));
+#endif
+}
+
+static void set_scmi_comm_resources(struct scpfw_config *cfg)
+{
+    unsigned int channel_index;
+    size_t i, j;
+    /* @cfg does not consider agent #0 this the reserved platform/server agent */
+    size_t scmi_agent_count = cfg->agent_count + 1;
+
+    scmi_agent_table = fwk_mm_calloc(scmi_agent_count,
+                                     sizeof(*scmi_agent_table));
+
+    scmi_service_elt = fwk_mm_calloc(scpfw_resource_counter.channel_count + 1,
+                                     sizeof(*scmi_service_elt));
+
+    msg_smt_elt = fwk_mm_calloc(scpfw_resource_counter.channel_count + 1,
+                                sizeof(*msg_smt_elt));
+    msg_smt_data = fwk_mm_calloc(scpfw_resource_counter.channel_count,
+                                 sizeof(*msg_smt_data));
+
+    tfm_mbx_elt = fwk_mm_calloc(scpfw_resource_counter.channel_count + 1,
+                                  sizeof(*tfm_mbx_elt));
+    tfm_mbx_data = fwk_mm_calloc(scpfw_resource_counter.channel_count,
+                                   sizeof(*tfm_mbx_data));
+
+    /* Set now the uniqnue scmi module instance configuration data */
+    scmi_data = (struct mod_scmi_config){
+        .agent_table = scmi_agent_table,
+        .agent_count = scmi_agent_count-1,
+        .protocol_count_max = 9,
+        .vendor_identifier = "STMicroelectronics",
+        .sub_vendor_identifier = "STMicroelectronics",
+    };
+
+    channel_index = 0;
+
+    for (i = 0; i < cfg->agent_count; i++) {
+        struct scpfw_agent_config *agent_cfg = cfg->agent_config + i;
+        size_t agent_index = i + 1;
+
+        scmi_agent_table[agent_index].type = SCMI_AGENT_TYPE_OSPM;
+        scmi_agent_table[agent_index].name = agent_cfg->name;
+
+        for (j = 0; j < agent_cfg->channel_count; j++) {
+            struct scpfw_channel_config *channel_cfg = agent_cfg->channel_config + j;
+            struct mod_scmi_service_config *service_data;
+
+            service_data = fwk_mm_calloc(1, sizeof(*service_data));
+            *service_data = (struct mod_scmi_service_config){
+                .transport_id = (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_MSG_SMT, channel_index),
+                .transport_api_id = (fwk_id_t)FWK_ID_API_INIT(FWK_MODULE_IDX_MSG_SMT,
+                                                              MOD_MSG_SMT_API_IDX_SCMI_TRANSPORT),
+                .scmi_agent_id = agent_cfg->agent_id,
+                .scmi_p2a_id = FWK_ID_NONE_INIT,
+            };
+
+            fwk_assert((service_data->scmi_agent_id == STM32MP25_AGENT_ID_M33_NS)||
+                       (service_data->scmi_agent_id == STM32MP25_AGENT_ID_CA35));
+
+            scmi_service_elt[channel_index].name = channel_cfg->name;
+            scmi_service_elt[channel_index].data = service_data;
+
+            msg_smt_elt[channel_index].name = channel_cfg->name;
+            msg_smt_elt[channel_index].data = (void *)(msg_smt_data + channel_index);
+
+            msg_smt_data[channel_index] = (struct mod_msg_smt_channel_config){
+                .type = MOD_MSG_SMT_CHANNEL_TYPE_REQUESTER,
+                .mailbox_size = 128,
+                .driver_id = (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_TFM_MBX,
+                                                           channel_index),
+                .driver_api_id = (fwk_id_t)FWK_ID_API_INIT(FWK_MODULE_IDX_TFM_MBX, 0),
+            };
+
+            tfm_mbx_elt[channel_index].name = channel_cfg->name;
+            tfm_mbx_elt[channel_index].data = (void *)(tfm_mbx_data + channel_index);
+
+            tfm_mbx_data[channel_index] = (struct mod_tfm_mbx_channel_config){
+                .driver_id = (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_MSG_SMT, channel_index),
+                .driver_api_id = (fwk_id_t)FWK_ID_API_INIT(FWK_MODULE_IDX_MSG_SMT,
+                                                           MOD_MSG_SMT_API_IDX_DRIVER_INPUT),
+            };
+
+            channel_index++;
+        }
+    }
+};
+
+static void set_resources(struct scpfw_config *cfg)
+{
+    size_t i, j, k;
+
+    for (i = 0; i < cfg->agent_count; i++) {
+        struct scpfw_agent_config *agent_cfg = cfg->agent_config + i;
+        size_t agent_index = i + 1;
+
+        if (agent_index != agent_cfg->agent_id) {
+
+            FWK_LOG_ERR("scpfw config expects agent ID is agent index");
+            panic();
+        }
+
+        for (j = 0; j < agent_cfg->channel_count; j++) {
+            struct scpfw_channel_config *channel_cfg = agent_cfg->channel_config + j;
+
+#ifdef CFG_SCPFW_MOD_SCMI_CLOCK
+            /* Add first SCMI clock. We will add later the clocks used for DVFS */
+            if (channel_cfg->clock_count) {
+                size_t clock_index = scpfw_resource_counter.clock_index;
+                struct mod_scmi_clock_device *dev = NULL;
+
+                /* Set SCMI clocks array for the SCMI agent */
+                dev = fwk_mm_calloc(channel_cfg->clock_count,
+                                    sizeof(struct mod_scmi_clock_device));
+
+                fwk_assert(!scmi_clk_agent_tbl[agent_index].device_table);
+                scmi_clk_agent_tbl[agent_index].device_count = channel_cfg->clock_count;
+                scmi_clk_agent_tbl[agent_index].device_table = dev;
+
+                /* Set clock and tfm/clock elements and config data */
+                for (k = 0; k < channel_cfg->clock_count; k++) {
+                    struct scmi_clock *clock_cfg = channel_cfg->clock + k;
+
+                    dev[k].element_id =
+                        (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_CLOCK, clock_index);
+
+                    tfm_clock_cfg[clock_index].clk = clock_cfg->clk;
+                    tfm_clock_cfg[clock_index].default_enabled = clock_cfg->enabled;
+
+                    tfm_clock_elt[clock_index].name = clock_cfg->name;
+                    tfm_clock_elt[clock_index].data = (void *)(tfm_clock_cfg + clock_index);
+
+                    clock_data[clock_index] = (struct mod_clock_dev_config){
+                        .driver_id = (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_TFM_CLOCK,
+                                                                   clock_index),
+                        .api_id = (fwk_id_t)FWK_ID_API_INIT(FWK_MODULE_IDX_TFM_CLOCK,
+                                                            0),
+                    };
+
+                    clock_elt[clock_index].name = clock_cfg->name;
+                    clock_elt[clock_index].data = (void *)(clock_data + clock_index);
+
+                    clock_index++;
+                }
+
+                scpfw_resource_counter.clock_index = clock_index;
+            }
+#endif
+
+#ifdef CFG_SCPFW_MOD_RESET_DOMAIN
+            if (channel_cfg->reset_count) {
+                struct mod_scmi_reset_domain_device *dev = NULL;
+                size_t reset_index = scpfw_resource_counter.reset_index;
+
+                /* Set SCMI reset domains array for the SCMI agent */
+                dev = fwk_mm_calloc(channel_cfg->reset_count, sizeof(*dev));
+
+                fwk_assert(!scmi_reset_agent_tbl[agent_index].device_table);
+                scmi_reset_agent_tbl[agent_index].agent_domain_count = channel_cfg->reset_count;
+                scmi_reset_agent_tbl[agent_index].device_table = dev;
+
+                /* Set reset_domain and tfm/reset elements and config data */
+                for (k = 0; k < channel_cfg->reset_count; k++) {
+                    struct scmi_reset *reset_cfg = channel_cfg->reset + k;
+
+                    dev[k].element_id =
+                        (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_RESET_DOMAIN,
+                                                      reset_index);
+
+                    tfm_reset_data[reset_index].rstctrl = reset_cfg->rstctrl;
+
+                    tfm_reset_elt[reset_index].name = reset_cfg->name;
+                    tfm_reset_elt[reset_index].data =
+                        (void *)(tfm_reset_data + reset_index);
+
+                    reset_data[reset_index] = (struct mod_reset_domain_dev_config){
+                        .driver_id =
+                            (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_TFM_RESET,
+                                                          reset_index),
+                        .driver_api_id =
+                            (fwk_id_t)FWK_ID_API_INIT(FWK_MODULE_IDX_TFM_RESET, 0),
+                        .modes = MOD_RESET_DOMAIN_AUTO_RESET |
+                                 MOD_RESET_DOMAIN_MODE_EXPLICIT_ASSERT |
+                                 MOD_RESET_DOMAIN_MODE_EXPLICIT_DEASSERT,
+                    };
+
+                    reset_elt[reset_index].name = reset_cfg->name;
+                    reset_elt[reset_index].data = (void *)(reset_data + reset_index);
+
+                    reset_index++;
+                }
+
+                scpfw_resource_counter.reset_index = reset_index;
+            }
+#endif
+
+#ifdef CFG_SCPFW_MOD_VOLTAGE_DOMAIN
+            if (channel_cfg->voltd_count) {
+                size_t regu_index = scpfw_resource_counter.regu_index;
+                struct mod_scmi_voltd_device *dev = NULL;
+
+                /* Set SCMI voltage domains array for the SCMI agent */
+                dev = fwk_mm_calloc(channel_cfg->voltd_count,
+                                    sizeof(struct mod_scmi_voltd_device));
+
+                fwk_assert(!scmi_voltd_agent_tbl[agent_index].device_table);
+                scmi_voltd_agent_tbl[agent_index].domain_count = channel_cfg->voltd_count;
+                scmi_voltd_agent_tbl[agent_index].device_table = dev;
+
+                /* Set voltage_domain and tfm/regu elements and config data */
+                for (k = 0; k < channel_cfg->voltd_count; k++) {
+                    struct scmi_voltd *voltd_cfg = channel_cfg->voltd + k;
+                    static const char reserved[] = "reserved";
+                    const char *name = NULL;
+
+                    if (voltd_cfg->dev) {
+                        name = voltd_cfg->dev->name;
+                    } else {
+                        name = reserved;
+                    }
+
+                    dev[regu_index].element_id =
+                       (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_VOLTAGE_DOMAIN, k);
+
+                    tfm_regu_elt[regu_index].name = name;
+                    tfm_regu_elt[regu_index].data = (void *)(tfm_regu_data + regu_index);
+                    tfm_regu_data[regu_index].dev = voltd_cfg->dev;
+                    tfm_regu_data[regu_index].default_enabled = voltd_cfg->enabled;
+
+                    voltd_data[regu_index].driver_id =
+                        (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_TFM_REGU_CONSUMER,
+                                                      regu_index);
+                    voltd_data[regu_index].api_id =
+                        (fwk_id_t)FWK_ID_API_INIT(FWK_MODULE_IDX_TFM_REGU_CONSUMER, 0);
+
+                    voltd_elt[regu_index].name = name;
+                    voltd_elt[regu_index].data = (void *)(voltd_data + regu_index);
+                    regu_index++;
+                }
+
+                scpfw_resource_counter.regu_index = regu_index;
+            }
+#endif
+        }
+    }
+}
+
+void scpfw_configure(struct scpfw_config *cfg)
+{
+    count_resources(cfg);
+    allocate_global_resources(cfg);
+    set_scmi_comm_resources(cfg);
+    set_resources(cfg);
+}
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_io.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_io.h
new file mode 100644
index 000000000..17105a268
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_io.h
@@ -0,0 +1,17 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef FMW_IO_H
+#define FMW_IO_H
+
+#include <fwk_id.h>
+#include <fwk_module_idx.h>
+
+#define FMW_IO_STDIN_ID FWK_ID_NONE
+#define FMW_IO_STDOUT_ID FWK_ID_ELEMENT(FWK_MODULE_IDX_TFM_CONSOLE, 0)
+
+#endif /* FMW_IO_H */
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_log.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_log.h
new file mode 100644
index 000000000..4e0bdd0ee
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/fmw_log.h
@@ -0,0 +1,16 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef FMW_LOG_H
+#define FMW_LOG_H
+
+/*
+ * Disable log buffering by setting buffer size to 0.
+ */
+#define FMW_LOG_BUFFER_SIZE 0
+
+#endif /* FMW_LOG_H */
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/gnu/stubs-soft.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/gnu/stubs-soft.h
new file mode 100644
index 000000000..e69de29bb
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/scmi_agents.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/scmi_agents.h
new file mode 100644
index 000000000..be74d9b91
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/include/scmi_agents.h
@@ -0,0 +1,13 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef SCMI_AGENTS_H
+#define SCMI_AGENTS_H
+#include <dt-bindings/scmi/stm32mp25-agents.h>
+
+
+#endif /* SCMI_AGENTS_H */
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/CMakeLists.txt b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/CMakeLists.txt
new file mode 100644
index 000000000..319139685
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/CMakeLists.txt
@@ -0,0 +1,25 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(${SCP_MODULE_TARGET} SCP_MODULE)
+
+target_include_directories(${SCP_MODULE_TARGET}
+                           PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")
+
+# These include directories are relative to SCP-firmware submodule position in
+# tfm-os tree
+target_include_directories(${SCP_MODULE_TARGET}
+                           PUBLIC ${SCP_TFM_PLAT_DIR}/native_driver/include
+                                  ${SCP_TFM_PLAT_DIR}/../devicetree/include
+                                  ${CMAKE_BINARY_DIR}/generated/devicetree/platform_s
+                                  ${SCP_TFM_DIR}/secure_fw/include)
+
+target_sources(${SCP_MODULE_TARGET}
+               PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/src/mod_tfm_clock.c")
+
+target_link_libraries(${SCP_MODULE_TARGET}
+                      PRIVATE module-clock)
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/Module.cmake b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/Module.cmake
new file mode 100644
index 000000000..760d695b5
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/Module.cmake
@@ -0,0 +1,9 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(SCP_MODULE "tfm-clock")
+set(SCP_MODULE_TARGET "module-tfm-clock")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/include/mod_tfm_clock.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/include/mod_tfm_clock.h
new file mode 100644
index 000000000..a5130247e
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/include/mod_tfm_clock.h
@@ -0,0 +1,32 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef MOD_TFM_CLOCK_H
+#define MOD_TFM_CLOCK_H
+
+#include <fwk_element.h>
+#include <fwk_macros.h>
+
+#include <clk.h>
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/*!
+ * \brief Platform clocks configuration.
+ */
+struct mod_tfm_clock_config {
+    /*! Clock name */
+    const char *name;
+    /*! Optee clock reference */
+    const struct clk *clk;
+    /*! default state of the clock */
+    bool default_enabled;
+};
+
+#endif /* MOD_TFM_CLOCK_H */
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/src/mod_tfm_clock.c b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/src/mod_tfm_clock.c
new file mode 100644
index 000000000..e84ff30c4
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_clock/src/mod_tfm_clock.c
@@ -0,0 +1,369 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022-2023, Linaro Limited and Contributors. All rights
+ * reserved.
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All
+ * rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Interface SCP-firmware clock module with stm32mp2 clock resources.
+ */
+
+#include <fwk_macros.h>
+#include <fwk_mm.h>
+#include <fwk_module.h>
+#include <fwk_log.h>
+
+#include <mod_clock.h>
+#include <mod_tfm_clock.h>
+
+#include <clk.h>
+
+#include <stdbool.h>
+#include <inttypes.h>
+
+
+#define MOD_NAME "[SCMI CLOCK] "
+
+/* Clock device context */
+struct tfm_clock_dev_ctx {
+    struct clk *clk;
+    bool enabled;
+};
+
+/* Clock module context */
+struct tfm_clock_module_ctx {
+    struct tfm_clock_dev_ctx *dev_ctx;
+    unsigned int dev_count;
+};
+
+static struct tfm_clock_module_ctx module_ctx;
+
+static struct tfm_clock_dev_ctx *elt_id_to_ctx(fwk_id_t dev_id)
+{
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return NULL;
+    }
+
+    return module_ctx.dev_ctx + fwk_id_get_element_idx(dev_id);
+}
+
+static bool is_exposed(struct tfm_clock_dev_ctx *ctx)
+{
+    return ctx->clk != NULL;
+}
+
+/*
+ * Clock driver API functions
+ */
+static int get_rate(fwk_id_t dev_id, uint64_t *rate)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+
+    if ((ctx == NULL) || (rate == NULL)) {
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        *rate = 0;
+        return FWK_SUCCESS;
+    }
+
+    *rate = clk_get_rate(ctx->clk);
+
+    FWK_LOG_DEBUG(
+        MOD_NAME "SCMI tfm_clock (%u/\"%s\"): clk_get_rate() = %" PRIu64,
+        fwk_id_get_element_idx(dev_id),
+        clk_get_name(ctx->clk),
+        *rate);
+
+    return FWK_SUCCESS;
+}
+
+static int set_state(fwk_id_t dev_id, enum mod_clock_state state)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    switch (state) {
+    case MOD_CLOCK_STATE_STOPPED:
+    case MOD_CLOCK_STATE_RUNNING:
+        break;
+    default:
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        if (state == MOD_CLOCK_STATE_STOPPED) {
+            return FWK_SUCCESS;
+        } else {
+            return FWK_E_ACCESS;
+        }
+    }
+
+    if (state == MOD_CLOCK_STATE_STOPPED) {
+        if (ctx->enabled) {
+            FWK_LOG_DEBUG(
+                MOD_NAME "SCMI tfm_clock (%u/\"%s\") disable",
+                fwk_id_get_element_idx(dev_id),
+                clk_get_name(ctx->clk));
+
+            clk_disable(ctx->clk);
+            ctx->enabled = false;
+        } else {
+            FWK_LOG_DEBUG(
+                MOD_NAME "SCMI tfm_clock (%u/\"%s\") is already OFF",
+                fwk_id_get_element_idx(dev_id),
+                clk_get_name(ctx->clk));
+        }
+    } else {
+        if (!ctx->enabled) {
+            FWK_LOG_DEBUG(
+                MOD_NAME "SCMI tfm_clock (%u/\"%s\") enable",
+                fwk_id_get_element_idx(dev_id),
+                clk_get_name(ctx->clk));
+
+            clk_enable(ctx->clk);
+            ctx->enabled = true;
+        } else {
+            FWK_LOG_DEBUG(
+                MOD_NAME "SCMI tfm_clock (%u/\"%s\") is already ON",
+                fwk_id_get_element_idx(dev_id),
+                clk_get_name(ctx->clk));
+        }
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int get_state(fwk_id_t dev_id, enum mod_clock_state *state)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+
+    if ((ctx == NULL) || (state == NULL)) {
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        *state = MOD_CLOCK_STATE_STOPPED;
+        return FWK_SUCCESS;
+    }
+
+    if (ctx->enabled) {
+        *state = MOD_CLOCK_STATE_RUNNING;
+    } else {
+        *state = MOD_CLOCK_STATE_STOPPED;
+    }
+
+    FWK_LOG_DEBUG(
+        MOD_NAME "SCMI tfm_clock (%u/\"%s\") is %s",
+        fwk_id_get_element_idx(dev_id),
+        clk_get_name(ctx->clk),
+        *state == MOD_CLOCK_STATE_STOPPED ? "off" : "on");
+
+    return FWK_SUCCESS;
+}
+
+static int get_range(fwk_id_t dev_id, struct mod_clock_range *range)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+    int res;
+    unsigned long min, max, step;
+
+    if ((ctx == NULL) || (range == NULL)) {
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        range->rate_type = MOD_CLOCK_RATE_TYPE_DISCRETE;
+        range->min = 0;
+        range->max = 0;
+        range->rate_count = 1;
+
+        return FWK_SUCCESS;
+    }
+
+    res = clk_get_rates_steps(ctx->clk, &min, &max, &step);
+
+    if (res) {
+        range->rate_type = MOD_CLOCK_RATE_TYPE_DISCRETE;
+        range->min = clk_get_rate(ctx->clk);
+        range->max = range->min;
+        range->rate_count = 1;
+    } else {
+        range->rate_type = MOD_CLOCK_RATE_TYPE_CONTINUOUS;
+        range->min = min;
+        range->max = max;
+        range->step = step;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int set_rate(fwk_id_t dev_id, uint64_t rate,
+                    enum mod_clock_round_mode round_mode)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+    int res;
+
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        return FWK_E_ACCESS;
+    }
+
+    res = clk_set_rate(ctx->clk, rate);
+    if (res) {
+        return FWK_E_SUPPORT;
+    }
+
+    FWK_LOG_DEBUG(
+        MOD_NAME "SCMI tfm_clock (%u/\"%s\"): rate = %" PRIu64,
+        fwk_id_get_element_idx(dev_id),
+        clk_get_name(ctx->clk),
+        rate);
+
+    return FWK_SUCCESS;
+}
+
+static int get_duty_cycle(fwk_id_t dev_id, uint32_t *num, uint32_t *den)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+    struct clk_duty duty = { 0 };
+    int res;
+
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        /* Return a dummy value to prevent an error trace */
+	*num = 1;
+	*den = 2;
+        return FWK_SUCCESS;
+    }
+
+    res = clk_get_duty_cycle(ctx->clk, &duty);
+
+    FWK_LOG_DEBUG(
+        MOD_NAME "SCMI tfm_clock (%u/\"%s\"): %" PRIx32,
+        fwk_id_get_element_idx(dev_id),
+        clk_get_name(ctx->clk),
+        res);
+
+    if (res) {
+        /* Assume a 50% duty cycle */
+        duty = (struct clk_duty){ .num = 1, .den = 2 };
+    } 
+    *num = duty.num;
+    *den = duty.den;
+
+    return FWK_SUCCESS;
+}
+
+static int get_rounded_rate(fwk_id_t dev_id, uint64_t rate,
+			    uint64_t *rounded_rate)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        return FWK_E_SUPPORT;
+    }
+
+    *rounded_rate = (uint64_t)clk_round_rate(ctx->clk, (unsigned long)rate);
+
+    return FWK_SUCCESS;
+}
+
+static int stub_process_power_transition(fwk_id_t dev_id, unsigned int state)
+{
+    return FWK_E_SUPPORT;
+}
+
+static int stub_pending_power_transition(fwk_id_t dev_id,
+                                         unsigned int current_state,
+                                         unsigned int next_state)
+{
+    return FWK_E_SUPPORT;
+}
+
+static const struct mod_clock_drv_api api_tfm_clock = {
+    .get_rate = get_rate,
+    .set_state = set_state,
+    .get_state = get_state,
+    .get_range = get_range,
+    .get_rate_from_index = NULL,
+    .set_rate = set_rate,
+    .get_duty_cycle = get_duty_cycle,
+    .round_rate = get_rounded_rate,
+    /* Not supported */
+    .process_power_transition = stub_process_power_transition,
+    .process_pending_power_transition = stub_pending_power_transition,
+};
+
+/*
+ * Framework handler functions
+ */
+
+static int tfm_clock_init(fwk_id_t module_id, unsigned int count,
+                            const void *data)
+{
+    if (count == 0) {
+        return FWK_SUCCESS;
+    }
+
+    module_ctx.dev_count = count;
+    module_ctx.dev_ctx = fwk_mm_calloc(count, sizeof(*module_ctx.dev_ctx));
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_clock_element_init(fwk_id_t element_id, unsigned int dev_count,
+                                    const void *data)
+{
+    struct tfm_clock_dev_ctx *ctx = elt_id_to_ctx(element_id);
+    const struct mod_tfm_clock_config *config = data;
+    int res;
+
+    ctx->clk = (struct clk *)config->clk;
+    if (ctx->clk) {
+        ctx->enabled = config->default_enabled;
+
+        if (ctx->enabled) {
+            res = clk_enable(ctx->clk);
+            if (res) {
+                return FWK_E_DEVICE;
+            }
+        }
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_clock_process_bind_request(fwk_id_t requester_id, fwk_id_t id,
+                                            fwk_id_t api_type, const void **api)
+{
+    *api = &api_tfm_clock;
+
+    return FWK_SUCCESS;
+}
+
+const struct fwk_module module_tfm_clock = {
+    .type = FWK_MODULE_TYPE_DRIVER,
+    .api_count = 1,
+    .event_count = 0,
+    .init = tfm_clock_init,
+    .element_init = tfm_clock_element_init,
+    .process_bind_request = tfm_clock_process_bind_request,
+};
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/CMakeLists.txt b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/CMakeLists.txt
new file mode 100644
index 000000000..21eee267d
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/CMakeLists.txt
@@ -0,0 +1,19 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(${SCP_MODULE_TARGET} SCP_MODULE)
+
+target_include_directories(${SCP_MODULE_TARGET}
+                           PUBLIC
+                                  ${CMAKE_CURRENT_SOURCE_DIR}/include
+                                  ${SCP_TFM_DIR}/platform/ext/common/)
+
+# These include directories are relative to SCP-firmware submodule position in
+# tfm tree
+
+target_sources(${SCP_MODULE_TARGET}
+               PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/src/mod_tfm_console.c")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/Module.cmake b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/Module.cmake
new file mode 100644
index 000000000..e661fd819
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/Module.cmake
@@ -0,0 +1,9 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(SCP_MODULE "tfm-console")
+set(SCP_MODULE_TARGET "module-tfm-console")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/src/mod_tfm_console.c b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/src/mod_tfm_console.c
new file mode 100644
index 000000000..1708cb8ab
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_console/src/mod_tfm_console.c
@@ -0,0 +1,93 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All rights
+ * reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Interface SCP-firmware console module with tfm console resources.
+ */
+
+#include <fwk_assert.h>
+#include <fwk_attributes.h>
+#include <fwk_event.h>
+#include <fwk_log.h>
+#include <fwk_mm.h>
+#include <fwk_module.h>
+#include <fwk_module_idx.h>
+#include <fwk_status.h>
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include <uart_stdout.h>
+static int mod_console_init(
+    fwk_id_t module_id,
+    unsigned int element_count,
+    const void *data)
+{
+    return FWK_SUCCESS;
+}
+
+static int mod_console_element_init(
+    fwk_id_t element_id,
+    unsigned int unused,
+    const void *data)
+{
+    return FWK_SUCCESS;
+}
+
+
+static int mod_console_io_open(const struct fwk_io_stream *stream)
+{
+    return FWK_SUCCESS;
+}
+
+static int mod_console_io_getch(
+    const struct fwk_io_stream *restrict stream,
+    char *restrict ch)
+{
+    *ch =  'A';
+
+    return FWK_SUCCESS;
+}
+
+static int mod_console_io_putch(const struct fwk_io_stream *stream, char ch)
+{
+    unsigned char c[2];
+    c[0] = (char)ch;
+    c[1] = 0;
+    stdio_output_string(c, 1);
+    return FWK_SUCCESS;
+}
+
+static int mod_console_close(const struct fwk_io_stream *stream)
+{
+    return FWK_SUCCESS;
+}
+
+const struct fwk_module module_tfm_console = {
+    .type = FWK_MODULE_TYPE_DRIVER,
+
+    .init = mod_console_init,
+    .element_init = mod_console_element_init,
+
+    .adapter =
+        (struct fwk_io_adapter){
+            .open = mod_console_io_open,
+            .getch = mod_console_io_getch,
+            .putch = mod_console_io_putch,
+            .close = mod_console_close,
+        },
+};
+
+struct fwk_module_config config_tfm_console = {
+    .elements = FWK_MODULE_STATIC_ELEMENTS({
+        [0] = {
+            .name = "printf",
+            .data = (void *)1
+        },
+        [1] = { 0 },
+    }),
+};
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/CMakeLists.txt b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/CMakeLists.txt
new file mode 100644
index 000000000..8af17e3b8
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/CMakeLists.txt
@@ -0,0 +1,27 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(${SCP_MODULE_TARGET} SCP_MODULE)
+
+target_include_directories(${SCP_MODULE_TARGET}
+                           PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
+
+# These include directories are relative to SCP-firmware submodule position in
+# tfm tree
+target_include_directories(${SCP_MODULE_TARGET}
+                           PUBLIC ${SCP_TFM_PLAT_DIR}/core/include/
+                                  {SCP_TFM_PLAT_DIR}/lib/libutils/ext/include/
+                                  ${SCP_TFM_PLAT_DIR}/lib/libutee/include/
+                                  ${SCP_TFM_DIR}/platform/ext/cmsis)
+
+target_link_libraries(${SCP_MODULE_TARGET}
+                      PRIVATE
+                              module-msg-smt
+                              module-scmi)
+
+target_sources(${SCP_MODULE_TARGET}
+               PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/src/mod_tfm_mbx.c")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/Module.cmake b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/Module.cmake
new file mode 100644
index 000000000..a3bca3f6f
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/Module.cmake
@@ -0,0 +1,10 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights
+# reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(SCP_MODULE "tfm-mbx")
+set(SCP_MODULE_TARGET "module-tfm-mbx")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/include/mod_tfm_mbx.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/include/mod_tfm_mbx.h
new file mode 100644
index 000000000..0330fa74c
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/include/mod_tfm_mbx.h
@@ -0,0 +1,67 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All rights
+ * reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     tfm mailbox buffer layer
+ */
+
+#ifndef MOD_TFM_MBX_H
+#define MOD_TFM_MBX_H
+
+/*!
+ * \brief Channel config.
+ */
+struct mod_tfm_mbx_channel_config {
+    /*! Identifier of the driver */
+    fwk_id_t driver_id;
+
+    /*! Identifier of the driver API to bind to */
+    fwk_id_t driver_api_id;
+};
+
+/*!
+ * \brief Interface to exchange message with OP-TEE
+ */
+
+/*!
+ * \brief Return the number of MBX devices registered in the paltform.
+ *
+ * \return Return the number of devices.
+ */
+int tfm_mbx_get_devices_count(void);
+
+/*!
+ * \brief Return the framwork ID related to the MBX device identifed by @id.
+ *
+ * \param id Device index.
+ *
+ * \return Return the framwork ID
+ */
+fwk_id_t tfm_mbx_get_device(unsigned int id);
+
+/*!
+ * \brief Signal an incoming SCMI message in a static shared memory.
+ *
+ * \param device_id MBX device ID.
+ */
+void tfm_mbx_signal_smt_message(fwk_id_t device_id);
+
+/*  */
+/*!
+ * \brief Signal an incoming SCMI message in an OP-TEE dynamic shared memory.
+ *
+ * \param device_id MBX device ID.
+ * \param in_buf Pointer to the request buffer.
+ * \param in_size Size of the request buffer.
+ * \param out_buf Pointer to the response buffer.
+ * \param out_size Size of the response buffer.
+ */
+void tfm_mbx_signal_msg_message(fwk_id_t device_id,
+                                  void *in_buf, size_t in_size,
+                                  void *out_buf, size_t *out_size);
+
+#endif /* MOD_TFM_MBX_H */
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c
new file mode 100644
index 000000000..062097506
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c
@@ -0,0 +1,258 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Handle the message coming from the tfm mailbox, invoke command transport
+ *     layer. This module remplaces the hardware mailbox device driver.
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include <fwk_id.h>
+#include <fwk_mm.h>
+#include <fwk_arch.h>
+#include <fwk_log.h>
+#include <fwk_module.h>
+#include <fwk_module_idx.h>
+#include <fwk_status.h>
+#include <mod_tfm_mbx.h>
+#include <mod_msg_smt.h>
+#include <cmsis_compiler.h>
+#define COMPILER_BARRIER() __ASM volatile("" : : : "memory")
+
+struct mutex {
+    uint8_t atomic;
+};
+
+static uint8_t atomic_add_u8(volatile uint8_t *valuePtr, int8_t delta)
+{
+    COMPILER_BARRIER();
+    uint8_t newValue;
+    do
+    {
+        newValue = __LDREXB(valuePtr) + delta;
+    } while (__STREXB(newValue, valuePtr));
+    COMPILER_BARRIER();
+    return newValue;
+}
+static void mutex_init(struct mutex *lock)
+{
+    lock->atomic = 0;
+}
+
+static void mutex_lock(struct mutex *lock)
+{
+    atomic_add_u8(&lock->atomic,1);
+    assert(lock->atomic == 1);
+}
+
+static void mutex_unlock(struct mutex *lock)
+{
+    atomic_add_u8(&lock->atomic,-1);
+    assert(lock->atomic == 0);
+}
+
+/* MBX device context */
+struct mbx_device_ctx {
+    /* Channel configuration data */
+    struct mod_tfm_mbx_channel_config *config;
+
+    fwk_id_t shmem_id;
+
+    union {
+        struct mod_tfm_smt_driver_input_api *smt;
+        struct mod_msg_smt_driver_input_api *msg;
+    } shmem_api;
+
+    size_t *shm_out_size;
+
+    struct mutex lock;
+};
+
+/* MBX context */
+struct mbx_ctx {
+    /* Table of device contexts */
+    struct mbx_device_ctx *device_ctx_table;
+
+    /* Number of devices in the device context table*/
+    unsigned int device_count;
+};
+
+static struct mbx_ctx mbx_ctx;
+
+#ifdef BUILD_HAS_MOD_MSG_SMT
+void tfm_mbx_signal_msg_message(fwk_id_t device_id, void *in_buf,
+                                  size_t in_size, void *out_buf,
+                                  size_t *out_size)
+{
+    struct mbx_device_ctx *device_ctx;
+    unsigned int device_idx = fwk_id_get_element_idx(device_id);
+
+    if (device_idx < mbx_ctx.device_count) {
+        device_ctx = &mbx_ctx.device_ctx_table[device_idx];
+
+        fwk_assert(fwk_id_get_module_idx(device_ctx->shmem_id) ==
+                   FWK_MODULE_IDX_MSG_SMT);
+
+
+        /* Lock the channel until the message has been processed */
+        mutex_lock(&device_ctx->lock);
+
+        device_ctx->shm_out_size = out_size;
+        device_ctx->shmem_api.msg->signal_message(device_ctx->shmem_id,
+                                                  in_buf, in_size,
+                                                  out_buf, *out_size);
+    } else {
+        fwk_unexpected();
+    }
+}
+#endif
+
+int tfm_mbx_get_devices_count(void)
+{
+    return mbx_ctx.device_count;
+}
+
+fwk_id_t tfm_mbx_get_device(unsigned int id)
+{
+    if (id >= mbx_ctx.device_count)
+        return (fwk_id_t)FWK_ID_NONE_INIT;
+
+    return (fwk_id_t)FWK_ID_ELEMENT_INIT(FWK_MODULE_IDX_TFM_MBX, id);
+}
+
+/*
+ * Mailbox module driver API
+ */
+
+#ifdef BUILD_HAS_MOD_MSG_SMT
+static int raise_shm_notification(fwk_id_t channel_id, size_t size)
+{
+    size_t idx = fwk_id_get_element_idx(channel_id);
+    struct mbx_device_ctx *channel_ctx = &mbx_ctx.device_ctx_table[idx];
+
+    *channel_ctx->shm_out_size = size;
+
+
+    /* Release the channel as the message has been processed */
+    mutex_unlock(&channel_ctx->lock);
+
+    return FWK_SUCCESS;
+}
+
+const struct mod_msg_smt_driver_ouput_api mbx_shm_api = {
+    .raise_notification = raise_shm_notification,
+};
+#endif
+/*
+ * Framework handlers
+ */
+
+static int mbx_init(fwk_id_t module_id, unsigned int device_count,
+                    const void *data)
+{
+
+    if (device_count == 0)
+        return FWK_E_PARAM;
+
+    mbx_ctx.device_ctx_table = fwk_mm_calloc(device_count,
+                                             sizeof(*mbx_ctx.device_ctx_table));
+    if (mbx_ctx.device_ctx_table == NULL) {
+        return FWK_E_NOMEM;
+    }
+
+    mbx_ctx.device_count = device_count;
+
+    return FWK_SUCCESS;
+}
+
+static int mbx_device_init(fwk_id_t device_id, unsigned int slot_count,
+                           const void *data)
+{
+    size_t elt_idx = fwk_id_get_element_idx(device_id);
+    struct mbx_device_ctx *device_ctx = &mbx_ctx.device_ctx_table[elt_idx];
+
+    device_ctx->config = (struct mod_tfm_mbx_channel_config*)data;
+
+    mutex_init(&device_ctx->lock);
+
+    return FWK_SUCCESS;
+}
+
+static int mbx_bind(fwk_id_t id, unsigned int round)
+{
+    int status;
+    struct mbx_device_ctx *device_ctx;
+
+    if (fwk_id_is_type(id, FWK_ID_TYPE_MODULE)) {
+        return FWK_SUCCESS;
+    }
+
+    if (round == 1) {
+        device_ctx = &mbx_ctx.device_ctx_table[fwk_id_get_element_idx(id)];
+
+        status = fwk_module_bind(device_ctx->config->driver_id,
+                                 device_ctx->config->driver_api_id,
+                                 &device_ctx->shmem_api);
+        if (status != FWK_SUCCESS) {
+            return status;
+        }
+
+        device_ctx->shmem_id = device_ctx->config->driver_id;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int mbx_process_bind_request(fwk_id_t source_id,
+                                    fwk_id_t target_id,
+                                    fwk_id_t api_id,
+                                    const void **api)
+{
+    size_t elt_idx;
+
+    if (!fwk_id_is_type(target_id, FWK_ID_TYPE_ELEMENT)) {
+        return FWK_E_ACCESS;
+    }
+
+    if (fwk_id_get_api_idx(api_id) != 0) {
+        return FWK_E_PARAM;
+    }
+
+    elt_idx = fwk_id_get_element_idx(target_id);
+    if (elt_idx >= mbx_ctx.device_count) {
+        return FWK_E_PARAM;
+    }
+
+    switch (fwk_id_get_module_idx(source_id)) {
+
+#ifdef BUILD_HAS_MOD_MSG_SMT
+    case FWK_MODULE_IDX_MSG_SMT:
+        *api = &mbx_shm_api;
+        break;
+#endif
+    default:
+        return FWK_E_PANIC;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int mbx_start(fwk_id_t id)
+{
+    return FWK_SUCCESS;
+}
+
+/* TFM_MBX module definition */
+const struct fwk_module module_tfm_mbx = {
+    .type = FWK_MODULE_TYPE_DRIVER,
+    .api_count = 1,
+    .init = mbx_init,
+    .element_init = mbx_device_init,
+    .bind = mbx_bind,
+    .start = mbx_start,
+    .process_bind_request = mbx_process_bind_request,
+};
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/CMakeLists.txt b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/CMakeLists.txt
new file mode 100644
index 000000000..d90bd6c9d
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/CMakeLists.txt
@@ -0,0 +1,25 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2023, Linaro Limited and Contributors. All rights reserved.
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(${SCP_MODULE_TARGET} SCP_MODULE)
+
+target_include_directories(${SCP_MODULE_TARGET}
+                           PUBLIC 
+                                  ${CMAKE_CURRENT_SOURCE_DIR}/include
+                                  ${SCP_SOURCE_DIR}/module/scmi/include
+                                  ${SCP_TFM_PLAT_DIR}/native_driver/include
+                                  ${SCP_TFM_PLAT_DIR}/../devicetree/include
+                                  ${SCP_TFM_PLAT_DIR}/native_driver
+                                  ${CMAKE_BINARY_DIR}/generated/devicetree/platform_s
+                                  ${SCP_TFM_PLAT_DIR}/../../include)
+
+
+target_sources(
+    ${SCP_MODULE_TARGET} PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/src/mod_tfm_regu_consumer.c")
+
+target_link_libraries(${SCP_MODULE_TARGET} PUBLIC module-voltage-domain)
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/Module.cmake b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/Module.cmake
new file mode 100644
index 000000000..daf632555
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/Module.cmake
@@ -0,0 +1,10 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2023, Linaro Limited and Contributors. All rights reserved.
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(SCP_MODULE "tfm-regu-consumer")
+set(SCP_MODULE_TARGET "module-tfm-regu-consumer")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/include/mod_tfm_regu_consumer.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/include/mod_tfm_regu_consumer.h
new file mode 100644
index 000000000..31250a6d7
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/include/mod_tfm_regu_consumer.h
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2020, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2024, STMicroelectronics
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef MOD_TFM_REGU_CONSUMER_H
+#define MOD_TFM_REGU_CONSUMER_H
+
+#include <fwk_element.h>
+#include <fwk_macros.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+struct scmi_server_regu_channel;
+struct rdev;
+
+/*!
+ * \brief Platform regulator configuration.
+ */
+struct mod_tfm_regu_consumer_dev_config {
+    const struct device *dev;
+    bool default_enabled;
+};
+
+#endif /* MOD_TFM_REGU_CONSUMER_H */
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/src/mod_tfm_regu_consumer.c b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/src/mod_tfm_regu_consumer.c
new file mode 100644
index 000000000..3e3bed72c
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_regu_consumer/src/mod_tfm_regu_consumer.c
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) 2017-2019, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2024, STMicroelectronics
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stddef.h>
+#include <regulator.h>
+#include <fwk_log.h>
+#include <fwk_macros.h>
+#include <fwk_mm.h>
+#include <fwk_module.h>
+#include <mod_scmi_std.h>
+#include <mod_tfm_regu_consumer.h>
+#include <mod_voltage_domain.h>
+
+#include <inttypes.h>
+#include <stdio.h>
+
+/* Device context */
+struct tfm_regu_consumer_dev_ctx {
+    struct device *dev;
+    bool enabled;
+};
+
+/* Module context */
+struct tfm_regu_consumer_ctx {
+    struct tfm_regu_consumer_dev_ctx *dev_ctx_table;
+    unsigned int dev_count;
+};
+
+/* A single instance handles all voltage regulators abstracted by regulator.h */
+static struct tfm_regu_consumer_ctx module_ctx;
+
+static char *regu_name(struct device *dev)
+{
+    if (dev)
+        return (char *)dev->name;
+
+    return NULL;
+}
+
+static int find_ctx(fwk_id_t dev_id,
+                    struct tfm_regu_consumer_dev_ctx **out_ctx)
+{
+    struct tfm_regu_consumer_dev_ctx *ctx;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id))
+        return FWK_E_PARAM;
+
+    if (!ctx->dev)
+        return FWK_E_ACCESS;
+
+    *out_ctx = ctx;
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_regu_consumer_get_config(fwk_id_t dev_id, uint8_t *mode_type,
+                                          uint8_t *mode_id)
+{
+    struct tfm_regu_consumer_dev_ctx *ctx;
+    int ret;
+
+    ret = find_ctx(dev_id, &ctx);
+    if (ret)
+        return ret;
+
+    *mode_type = MOD_VOLTD_MODE_TYPE_ARCH;
+
+    if (ctx->enabled)
+        *mode_id = MOD_VOLTD_MODE_ID_ON;
+    else
+        *mode_id = MOD_VOLTD_MODE_ID_OFF;
+
+    FWK_LOG_INFO("SCMI voltd %u: get config PMIC %s = %#"PRIx8", %#"PRIx8,
+              fwk_id_get_element_idx(dev_id), regu_name(ctx->dev), *mode_type,
+              *mode_id);
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_regu_consumer_set_config(fwk_id_t dev_id, uint8_t mode_type,
+                                          uint8_t mode_id)
+{
+    struct tfm_regu_consumer_dev_ctx *ctx;
+    int ret;
+
+    ret = find_ctx(dev_id, &ctx);
+    if (ret)
+        return ret;
+
+    if (mode_type != MOD_VOLTD_MODE_TYPE_ARCH) {
+        return FWK_E_PARAM;
+    }
+
+    switch (mode_id) {
+    case MOD_VOLTD_MODE_ID_ON:
+        if (!ctx->enabled) {
+            if (regulator_enable(ctx->dev))
+                return FWK_E_DEVICE;
+
+            ctx->enabled = true;
+        }
+        break;
+
+    case MOD_VOLTD_MODE_ID_OFF:
+        if (ctx->enabled) {
+            if (regulator_disable(ctx->dev))
+                return FWK_E_DEVICE;
+
+            ctx->enabled = false;
+        }
+        break;
+
+    default:
+        return FWK_E_PARAM;
+    }
+
+    FWK_LOG_INFO("SCMI voltd %u: set config PMIC %s to type %#"PRIx8" / mode %#"PRIx8,
+              fwk_id_get_element_idx(dev_id), regu_name(ctx->dev), mode_type,
+              mode_id);
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_regu_consumer_get_level(fwk_id_t dev_id, int32_t *level_uv)
+{
+    struct tfm_regu_consumer_dev_ctx *ctx;
+    int32_t level_mv;
+    int ret;
+
+    ret = find_ctx(dev_id, &ctx);
+    if (ret)
+        return ret;
+
+    if (regulator_get_voltage(ctx->dev, &level_mv))
+        return FWK_E_PANIC;
+
+    *level_uv = (int)level_mv * 1000;
+
+    FWK_LOG_INFO("SCMI voltd %u: get level PMIC %s = %ld",
+              fwk_id_get_element_idx(dev_id), regu_name(ctx->dev), *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_regu_consumer_set_level(fwk_id_t dev_id, int32_t level_uv)
+{
+    struct tfm_regu_consumer_dev_ctx *ctx = NULL;
+    int ret = FWK_E_PANIC;
+
+    if (level_uv / 1000 > UINT16_MAX) {
+        FWK_LOG_ERR("Volatge level too high (mV shall fit in 16 bits)");
+        return FWK_E_PARAM;
+    }
+
+    ret = find_ctx(dev_id, &ctx);
+    if (ret)
+        return ret;
+
+    if (regulator_set_voltage(ctx->dev, level_uv / 1000, level_uv/1000))
+        return FWK_E_DEVICE;
+
+    FWK_LOG_INFO("SCMI voltd %u: set level PMIC %s to %ld",
+              fwk_id_get_element_idx(dev_id), regu_name(ctx->dev), level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_regu_consumer_get_info(fwk_id_t dev_id,
+                                        struct mod_voltd_info *info)
+{
+    struct tfm_regu_consumer_dev_ctx *ctx;
+    int full_count;
+    int ret;
+    unsigned int min_idx, max_idx;
+    int32_t volt_uv[2];
+
+    ret = find_ctx(dev_id, &ctx);
+    if (ret == FWK_E_ACCESS) {
+        static const char reserved[] = "reserved";
+
+        memset(info, 0, sizeof(*info));
+        info->name = reserved;
+        info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+        info->level_range.level_count = 1;
+        info->level_range.min_uv = 0;
+        info->level_range.max_uv = 0;
+
+        return FWK_SUCCESS;
+    } else if (ret != FWK_SUCCESS) {
+        return ret;
+    }
+
+    full_count = regulator_count_voltages(ctx->dev);
+
+    if (full_count == 0)
+       return FWK_E_SUPPORT;
+    min_idx = 0;
+    max_idx = full_count-1;
+    if (regulator_list_voltage(ctx->dev, min_idx, &volt_uv[0])) {
+        return FWK_E_SUPPORT;
+    }
+    if (regulator_list_voltage(ctx->dev, max_idx, &volt_uv[1])) {
+        return FWK_E_SUPPORT;
+    }
+
+
+    memset(info, 0, sizeof(*info));
+    info->name = regu_name(ctx->dev);
+    info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+    info->level_range.level_count = full_count;
+    info->level_range.min_uv = volt_uv[0]*1000;
+    info->level_range.max_uv = volt_uv[1]*1000;
+
+    FWK_LOG_INFO("SCMI voltd %u: get_info PMIC %s, range [%ld %ld]",
+              fwk_id_get_element_idx(dev_id), regu_name(ctx->dev),
+              info->level_range.min_uv, info->level_range.max_uv);
+
+    return FWK_SUCCESS;
+}
+
+static const struct mod_voltd_drv_api api_tfm_regu = {
+    .get_level = tfm_regu_consumer_get_level,
+    .set_level = tfm_regu_consumer_set_level,
+    .set_config = tfm_regu_consumer_set_config,
+    .get_config = tfm_regu_consumer_get_config,
+    .get_info = tfm_regu_consumer_get_info,
+    .get_level_from_index = NULL,
+};
+
+/*
+ * Framework handler functions
+ */
+
+static int tfm_regu_consumer_init(fwk_id_t module_id,
+                                    unsigned int element_count,
+                                    const void *data)
+{
+    module_ctx.dev_count = element_count;
+
+    if (element_count)
+        module_ctx.dev_ctx_table =
+            fwk_mm_calloc(element_count, sizeof(*module_ctx.dev_ctx_table));
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_regu_consumer_element_init(fwk_id_t element_id,
+                                            unsigned int unused,
+                                            const void *data)
+{
+    const struct mod_tfm_regu_consumer_dev_config *dev_config = data;
+    struct tfm_regu_consumer_dev_ctx *ctx;
+
+    if (!fwk_module_is_valid_element_id(element_id))
+        return FWK_E_PANIC;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(element_id);
+
+    ctx->dev = (struct device *)dev_config->dev;
+    ctx->enabled = dev_config->default_enabled;
+
+    if (ctx->enabled) {
+        if (!ctx->dev)
+            return FWK_E_PARAM;
+        if (regulator_enable(ctx->dev))
+            return FWK_E_DEVICE;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_regu_consumer_process_bind_request(fwk_id_t requester_id,
+                                                    fwk_id_t target_id,
+                                                    fwk_id_t api_type,
+                                                    const void **api)
+{
+    *api = &api_tfm_regu;
+
+    return FWK_SUCCESS;
+}
+
+const struct fwk_module module_tfm_regu_consumer = {
+    .type = FWK_MODULE_TYPE_DRIVER,
+    .api_count = 1,
+    .init = tfm_regu_consumer_init,
+    .element_init = tfm_regu_consumer_element_init,
+    .process_bind_request = tfm_regu_consumer_process_bind_request,
+};
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/CMakeLists.txt b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/CMakeLists.txt
new file mode 100644
index 000000000..541d3fa96
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/CMakeLists.txt
@@ -0,0 +1,24 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(${SCP_MODULE_TARGET} SCP_MODULE)
+
+target_include_directories(${SCP_MODULE_TARGET}
+                           PUBLIC
+                           ${CMAKE_CURRENT_SOURCE_DIR}/include
+                           ${SCP_TFM_PLAT_DIR}/core/include/
+                           ${SCP_TFM_PLAT_DIR}/lib/libutils/ext/include/
+                           ${SCP_TFM_PLAT_DIR}/lib/libutee/include/
+                          ${SCP_TFM_DIR}/platform/ext/cmsis)
+
+
+target_sources(${SCP_MODULE_TARGET}
+               PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/src/mod_tfm_reset.c")
+
+target_link_libraries(${SCP_MODULE_TARGET}
+                      PRIVATE module-reset-domain)
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/Module.cmake b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/Module.cmake
new file mode 100644
index 000000000..8500d975e
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/Module.cmake
@@ -0,0 +1,10 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+# Copyright (c) 2024, STMicroelectronics and the Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(SCP_MODULE "tfm-reset")
+set(SCP_MODULE_TARGET "module-tfm-reset")
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/include/mod_tfm_reset.h b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/include/mod_tfm_reset.h
new file mode 100644
index 000000000..f85a61d31
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/include/mod_tfm_reset.h
@@ -0,0 +1,23 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022, Linaro Limited and Contributors. All rights reserved.
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All rights
+ * reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef MOD_TFM_RESET_H
+#define MOD_TFM_RESET_H
+
+/*!
+ * \brief Platform reset domain configuration.
+ */
+struct mod_tfm_reset_dev_config {
+    /*! Optee reset reference */
+    const struct reset_control *rstctrl;
+    /*! Reset line name */
+    const char *name;
+};
+
+#endif /* MOD_TFM_RESET_H */
diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/src/mod_tfm_reset.c b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/src/mod_tfm_reset.c
new file mode 100644
index 000000000..ee3ff6ec4
--- /dev/null
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_reset/src/mod_tfm_reset.c
@@ -0,0 +1,172 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2022-2023, Linaro Limited and Contributors. All rights
+ * reserved.
+ * Copyright (c) 2024, STMicroelectronics and the Contributors. All
+ * rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Interface SCP-firmware reset module with tfm reset resources.
+ */
+
+#include <fwk_macros.h>
+#include <fwk_mm.h>
+#include <fwk_module.h>
+#include <fwk_log.h>
+#include <inttypes.h>
+#include <mod_tfm_reset.h>
+#include <mod_reset_domain.h>
+
+#include <reset.h>
+
+#include <stdbool.h>
+
+#define MOD_NAME "[SCMI RESET] "
+
+#define TIMEOUT_US_1MS 1000
+
+/* Device context */
+struct tfm_reset_dev_ctx {
+    struct reset_control *rstctrl;
+};
+
+/* Module context */
+struct tfm_reset_ctx {
+    struct tfm_reset_dev_ctx *dev_ctx_table;
+    unsigned int dev_count;
+};
+
+static struct tfm_reset_ctx module_ctx;
+
+static bool is_exposed(struct tfm_reset_dev_ctx *ctx)
+{
+    return ctx->rstctrl != NULL;
+}
+
+/*
+ * Driver API functions
+ */
+static int reset_set_state(fwk_id_t dev_id, enum mod_reset_domain_mode mode,
+                           uint32_t reset_state, uintptr_t cookie)
+{
+    struct tfm_reset_dev_ctx *ctx = NULL;
+    int status = FWK_SUCCESS;
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    ctx = &module_ctx.dev_ctx_table[fwk_id_get_element_idx(dev_id)];
+
+    if (!is_exposed(ctx)) {
+        return FWK_E_ACCESS;
+    }
+
+    /* Whatever the reset_state set, we consider a unique context loss mode */
+    if (reset_state) {
+        FWK_LOG_INFO(MOD_NAME "Override requested SCMI reset state %#"PRIx32,
+            reset_state);
+    }
+
+    switch (mode) {
+    case MOD_RESET_DOMAIN_MODE_EXPLICIT_ASSERT:
+        FWK_LOG_DEBUG(
+            MOD_NAME "SCMI reset %u: assert(%s%d)",
+            fwk_id_get_element_idx(dev_id),
+            ctx->rstctrl->dev->name, ctx->rstctrl->id);
+
+        if (reset_control_assert(ctx->rstctrl)) {
+            status = FWK_E_DEVICE;
+        }
+        break;
+    case MOD_RESET_DOMAIN_MODE_EXPLICIT_DEASSERT:
+        FWK_LOG_DEBUG(
+            MOD_NAME "SCMI reset %u: deassert(%s%d)",
+            fwk_id_get_element_idx(dev_id),
+            ctx->rstctrl->dev->name, ctx->rstctrl->id);
+
+        if (reset_control_deassert(ctx->rstctrl)) {
+            status = FWK_E_DEVICE;
+        }
+        break;
+    case MOD_RESET_DOMAIN_AUTO_RESET:
+        FWK_LOG_DEBUG(
+            MOD_NAME "SCMI reset %u: cycle(%s)",
+            fwk_id_get_element_idx(dev_id),
+            fwk_id_get_element_idx(dev_id),
+            ctx->rstctrl->dev->name, ctx->rstctrl->id);
+;
+
+        if (reset_control_reset(ctx->rstctrl)) {
+            status = FWK_E_TIMEOUT;
+        }
+        if (reset_control_reset(ctx->rstctrl)) {
+            status = FWK_E_TIMEOUT;
+        }
+        break;
+    default:
+        return FWK_E_PARAM;
+    }
+
+    return status;
+}
+
+static const struct mod_reset_domain_drv_api api_tfm_reset = {
+    .set_reset_state = reset_set_state,
+};
+
+/*
+ * Framework handler functions
+ */
+
+static int tfm_reset_init(fwk_id_t module_id, unsigned int element_count,
+                            const void *data)
+{
+    module_ctx.dev_count = element_count;
+
+    if (element_count == 0) {
+        return FWK_SUCCESS;
+    }
+
+    module_ctx.dev_ctx_table = fwk_mm_calloc(element_count,
+                                             sizeof(struct tfm_reset_dev_ctx));
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_reset_element_init(fwk_id_t element_id,
+                                    unsigned int unused __unused,
+                                    const void *data)
+{
+    struct tfm_reset_dev_ctx *ctx = NULL;
+    const struct mod_tfm_reset_dev_config *config = NULL;
+
+    if (!fwk_module_is_valid_element_id(element_id)) {
+        return FWK_E_PARAM;
+    }
+
+    ctx = &module_ctx.dev_ctx_table[fwk_id_get_element_idx(element_id)];
+    config = (const struct mod_tfm_reset_dev_config *)data;
+
+    ctx->rstctrl = (struct reset_control *)config->rstctrl;
+
+    return FWK_SUCCESS;
+}
+
+static int tfm_reset_process_bind_request(fwk_id_t requester_id, fwk_id_t id,
+                                            fwk_id_t api_type, const void **api)
+{
+    *api = &api_tfm_reset;
+
+    return FWK_SUCCESS;
+}
+
+const struct fwk_module module_tfm_reset = {
+    .type = FWK_MODULE_TYPE_DRIVER,
+    .api_count = 1,
+    .init = tfm_reset_init,
+    .element_init = tfm_reset_element_init,
+    .process_bind_request = tfm_reset_process_bind_request,
+};
-- 
2.47.2

